File: Particle_Tracker/Julia/Field.jl
using StaticArrays

# export Field, weightedforce


struct Field
    field::Array{Float64, 4}

    function Field(nx::UInt64, ny::UInt64, nz::UInt64, f::Function)
        new(f(nx, ny, nz))
    end
end


function weightedforce(field::Field, pos::MVector{3, Float64},
                       nx::UInt64, ny::UInt64, nz::UInt64,
                       inv_step_size::SVector{3, Float64})::MVector{3, Float64}
    # This algorithm was written dependent on a language that uses a 0-base
    # index. Julia is a 1-based index system. As such, 1 needs to be added to
    # ii in order for its index to be in bounds for a Julia array.
    qoverm = one(Float64)

    localforce = zeros(MVector{3, Float64})
    volumes = zeros(MArray{Tuple{2, 2, 2}, Float64})

    weight = copy(pos)
    weight .*= inv_step_size
    ii = map(x -> round(Int64, x, RoundDown), weight) # Add one when necessary
    weight .-= map(x -> convert(Float64, x), ii)

    for i = 1:size(volumes)[1], j = 1:size(volumes)[2], k = 1:size(volumes)[3]
        # Subtract 1 from i, j, k so they work with the calculation.
        @inbounds volumes[i, j, k] = abs(((one(Float64) - (i - 1)) - weight[1]) * ((one(Float64) - (j - 1)) - weight[2]) * ((one(Float64) - (k - 1)) - weight[3]))
    end

    for i = 1:size(volumes)[1], j = 1:size(volumes)[2], k = 1:size(volumes)[3]
        # Adding 1 to ii and subtracting 1 from i, j, k cancel out here
        i0 = ii[1] + i
        i1 = ii[2] + j
        i2 = ii[3] + k
        # Adjust the index value checks to match Julia array indexing
        if i0 <= nx && i1 <= ny && i2 <= nz &&
           i0 > 0 && i1 > 0 && i2 > 0
            @. @inbounds localforce += volumes[i, j, k] * field.field[:, i0, i1, i2]
        else
            println("WARNING: Evaluating field outside specified domain. (It treats this as if there is no field at this position.)")
        end
    end

    return localforce .* qoverm
end

File: Particle_Tracker/Python/Field.py
import numpy as np
from FrozenClass import FrozenClass

ndims = np.uint8(3)


class Field(FrozenClass):
    def __init__(self, nx, ny, nz, strength, rmin, rmax, f):
        self.field = f(nx, ny, nz, ndims, strength, rmin, rmax)
        return

    def set_field(self, f):
        msg = ("incompatible dimensions for field\n"
               "(dimension must be %r)")
        if f.shape != self.field.shape:
            raise ValueError(msg % (self.field.shape,))

        self.field = f
        return

    def get_field(self):
        return self.field

    def Weighted_Force(self, pos, nx, ny, nz, inv_step_size):
        # linear interpolation in 3d

        # variables to check whether any of the particle still lies
        # within the field domain
        domain_count = 0
        domain_status = True

        # charge to mass ratio - normalized to |e|/m_e
        qoverm = np.float64(-1.0)

        local_force = np.zeros(ndims, dtype=np.float64)

        volumes = np.zeros((2, 2, 2), dtype=np.float64)

        weight = np.copy(pos)
        weight *= inv_step_size
        ii = weight.astype(np.int64)
        weight -= ii.astype(np.float64)

        for index, in np.ndenumerate(volumes):
            volumes[index] = np.absolute(((np.float64(1.0) - index[0]) - weight[0]) * ((np.float64(1.0) - index[1]) - weight[1]) * ((np.float64(1.0) - index[2]) - weight[2]), dtype=np.float64)

        for index, in np.ndenumerate(volumes):
            i0 = ii[0] + index[0]
            i1 = ii[1] + index[1]
            i2 = ii[2] + index[2]
            if ((i0 < nx) and (i1 < ny) and (i2 < nz) and
                    (i0 >= 0) and (i1 >= 0) and (i2 >= 0)):
                local_force += volumes[index] * self.field[:, i0, i1, i2]
            else:
                domain_count += 1
                # print "WARNING: Evaluating field outside specified domain. (It treats this as if there is no field at this position.)"

        if domain_count == volumes.size:
            domain_status = False

        return local_force * qoverm, domain_status
